### PR TITLE
GPU acceleration by parallel diagonalization

### PR DESCRIPTION
**Pull request by anjohan**
This branch includes the batched diagonalization in the phonon and electronWannier Hamiltonians, which are called from the bandstructure constructor.

Note: Most of the eigenvector unit tests have been disabled. Everything "works" in the sense that the thermal conductivity comes out correct, and there is freedom of both phase and degenerate subspace linear combinations, making the eigenvectors hard to test.

Note: Only properly tested on our limited test cases (silicon etc.).

Future enhancements that may be necessary:
- The same parallelization in buildAsPostProcessing
- The same parallelization in buildOnTheFly
- More careful batching in case of memory issues/limitation
- More comments

### CODE REVIEW

**Review by cepellotti — APPROVED** [review]

File: src/harmonic/electron_h0_wannier.cpp [hunk 1]
```diff
@@ -375,4 +497,435 @@ void ElectronH0Wannier::addShiftedVectors(Eigen::Tensor<double,3> degeneracyShif
...
+    double newMemory = getDeviceMemoryUsage();
+    kokkosDeviceMemory->addDeviceMemoryUsage(newMemory);
+  }
 }
+
+
+std::tuple<std::vector<Eigen::VectorXd>,
+           std::vector<Eigen::MatrixXcd>,
+           std::vector<Eigen::Tensor<std::complex<double>, 3>>>
+ElectronH0Wannier::batchedDiagonalizeWithVelocities(
```
Comment (jcoulter12) [inline src/harmonic/electron_h0_wannier.cpp:547]: Can I have a comment on the indexing here please?

File: src/harmonic/electron_h0_wannier.cpp [hunk 2]
```diff
@@ -148,154 +236,188 @@ ElectronH0Wannier::diagonalizeVelocity(Point &point) {
...
-    // and to be safe, we reimpose hermiticity
-    der = 0.5 * (der + der.adjoint());
-
-    // now we rotate in the basis of the eigenvectors at q.
-    der = eigenvectors.adjoint() * der * eigenvectors;
-
-    for (int ib1 = 0; ib1 < numWannier; ib1++) {
-      for (int ib2 = 0; ib2 < numWannier; ib2++) {
-        velocity(ib1, ib2, i) = der(ib1, ib2);
+    DoubleView2D cartesianWavevectors_d("el_cartWav_d", numK, 3);
```
Comment (jcoulter12) [inline src/harmonic/electron_h0_wannier.cpp:270]: Could we have a comment on what this block is for?

File: src/harmonic/electron_h0_wannier.cpp [hunk 3]
```diff
@@ -375,4 +497,435 @@ void ElectronH0Wannier::addShiftedVectors(Eigen::Tensor<double,3> degeneracyShif
...
+  );
+
+  StridedComplexView3D hamiltonians("hamiltonians", Hlayout);
+  Kokkos::complex<double> complexI(0.0, 1.0);
+
+  auto bravaisVectors_d = this->bravaisVectors_d;
+  auto vectorsDegeneracies_d = this->vectorsDegeneracies_d;
+  auto h0R_d = this->h0R_d;
+
+  // not sure what this does
```
Comment (jcoulter12) [inline src/harmonic/electron_h0_wannier.cpp]: @cepellotti maybe you know what this does? :)

File: src/harmonic/electron_h0_wannier.cpp [hunk 4]
```diff
@@ -375,4 +497,435 @@ void ElectronH0Wannier::addShiftedVectors(Eigen::Tensor<double,3> degeneracyShif
...
+    const DoubleView2D &cartesianCoordinates) {
+
+  int numWannier = this->numWannier; // Kokkos quirkyness
+
+  int numK = cartesianCoordinates.extent(0);
+
+  double delta = 1.0e-8;
+  double threshold = 0.000001 / energyRyToEv;// = 1 micro-eV
+
+  // prepare all the wavevectors at which we need the hamiltonian
```
Comment (jcoulter12) [inline src/harmonic/electron_h0_wannier.cpp:809]: @anjohan Again maybe a comment about indexing (if it's not exactly the same as somewhere else in this part of the code)?

File: src/harmonic/phonon_h0.cpp [hunk 5]
```diff
@@ -561,6 +696,15 @@ void PhononH0::shortRangeTerm(Eigen::Tensor<std::complex<double>, 4> &dyn,
       }
     }
   }
+    //for (int nb = 0; nb < numAtoms; nb++) {
```
Comment (jcoulter12) [inline src/harmonic/phonon_h0.cpp:699]: @anjohan there's a bunch of commented print statements in this function, do we want to keep them still for some reason?

Comment (jcoulter12) [reply src/harmonic/phonon_h0.cpp:699]: We discussed and decided to keep them.

**Comment by jcoulter12** [comment]
Ok, other than the few places I noted on the PR that it would be good to have another comment, and the note that we should check GaN band structure since the long range correction was edited, this looks good to me!

Thanks @cepellotti and @anjohan for all the work on this PR!

**Comment by anjohan** [comment]
I've added some more comments.

The electronWannierTransport has also been re-checked for GaN and shows good agreement with the develop branch.

**Pull request merged**